src/coreclr/src/jit/importer.cpp

```diff
@@ -10278,6 +10278,23 @@ GenTree* Compiler::impOptimizeCastClassOrIsInst(GenTree* op1, CORINFO_RESOLVED_T
         }
         else
         {
+            if (!isCastClass && op1->IsBoxedValue())
+            {
+                TypeHandle fromType(fromClass);
+                TypeHandle toType(toClass);
+
+                if (Nullable::IsNullableForTypeNoGC(fromType, toType.GetMethodTable()))
```

Comment:
I am pretty sure you can't call Nullable from here, it's in the vm and this code is in the jit. The problem you are trying to solve as far as I understand is [here](https://github.com/dotnet/runtime/blob/fd181c0210c263e40ce8b908bdd521d8f3cc284e/src/coreclr/src/vm/jitinterface.cpp#L4467-L4471).

Reply:
Right, there is a "jit interface" boundary between jit and runtime, specified in corinfo.h; the jit can only access the runtime via the methods in that interface. There are multiple implementations of this interface, eg one in the runtime proper used when jitting, another in crossgen used when prejitting, yet another in crossgen2.

So lyou should instead update the jit interface method `compareTypesForCast` in the runtime to call `Nullable::IsNullableForType`, returning a new result code if this is true, and use that result code to drive the transformation above, and then make similar changes in crossgen2 (crossgen's jit interface is built on top of the runtime version, so will pick up the runtime change). Also you would need to audit the other places the jit calls `compareTypesForCast` to see if any of those need changes.

You could perhaps try and check conditions 1 and 2 on the jit side; this might be doable, but would be a bit roundabout and less efficient.

Reply:
Got it, thank you! Where can I find more info how CLR is organized?

Reply:
Try the [book of the runtime](https://github.com/dotnet/runtime/tree/master/docs/design/coreclr/botr).

Reply:
Read it, but probably missed information about boundaries...

Reply:
There probably needs to be a "big picture" architecture view in there somewhere showing the relationships between the major parts of the system, and the discussing the rationale for the boundaries.

Reply:
https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/ryujit-overview.md#execution-environment-and-external-interface ?